### PR TITLE
PBM-773: distributed transactions

### PR DESCRIPTION
**Pull request by dAdAbird**
Distributed transactions always consist of at least two ops - `prepare`
(with txn ops) and `commitTransaction` or abortTransaction.
Although commitTimestamp for the same txn on different shards always the same,
the op with `commitTransaction` may have different `ts` (op timestamp). Fo example:
```
rs1: { "lsid" : { "id" : UUID("f7ad867d-f9a4-444c-bf5f-7185e7038d6e"), ... },
	"o" : { "commitTransaction" : 1, "commitTimestamp" : Timestamp(1644410656, 6) },
	"ts" : Timestamp(1644410656, 9), ... }
rs2: { "lsid" : { "id" : UUID("f7ad867d-f9a4-444c-bf5f-7185e7038d6e"), ... },
	"o" : { "commitTransaction" : 1, "commitTimestamp" : Timestamp(1644410656, 6) },
	"ts" : Timestamp(1644410656, 8), ... }
```
Since we sync backup/restore across shards by `ts` (`opTime`), artifacts of such transaction
 would be visible on shard `rs2` and won't appear on `rs1` given the restore time is `(1644410656, 8)`.
To avoid that we have to check if a distributed transaction was committed on all
participated shards before committing such transaction.

Oplog applier:
- Encountering `prepare` statement `handleTxnOp` would send such txn to the caller.
- Bumping into `commitTransaction` it will send txn again with respective state and
	 commit time. And waits for the response from the caller with either "commit" or "abort" state.
- On "abort" transactions buffer will be purged, hence all ops are discarded.

The restore routine after calling applier will run `distTxnChecker` which:
- Receiving distributed transactions from the oplog applier, will add set it to the shards restore meta.
- If txn's state is `commit` it will wait from all of the rest of the shards for:
     - either transaction committed on the shard (have `commitTransaction`);
     - or shard didn't participate in the transaction (more on that below);
     - or shard participated in the txn (have prepare ops) but no `commitTransaction` by the end of the oplog.
  
 If any of the shards encounters the latter - the transaction is sent back to the applier as aborted.
	  Otherwise - committed.

By looking at just transactions in the oplog we can't tell which shards were participating in it. So given that
starting `opTime` of the transaction is the same on all shards, we can assume that if some shard(s) oplog applier
observed greater than the txn's `opTime` and hadn't seen such txn - it wasn't part of this transaction at all. To
communicate that, each agent runs `checkWaitingTxns` which in turn periodically checks restore metadata and sees
any new (unobserved before) waiting for the transaction, posts last observed opTime. We go with `checkWaitingTxns`
instead of just updating each observed `opTime`  since the latter would add an extra 1 write to each oplog op on
sharded clusters even if there are no dist txns at all.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | failed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.2 | failed |
| Non-sharded replicaset mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.2 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Single-node replicaset mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.0 | failed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.4 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Restore on new cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 4.0 | failed |
| Restore on new cluster mongodb 4.2 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.4 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | passed |
| Sharded cluster mongodb 4.4 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 5.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | passed |
| Restore on new cluster mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.2 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | failed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Restore on new cluster mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 5.0 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | passed |
| Sharded cluster mongodb 4.4 | passed |
| Sharded cluster mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.4 | failed |
| Restore on new cluster mongodb 5.0 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 5.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | passed |
| Sharded cluster mongodb 4.4 | passed |
| Sharded cluster mongodb 5.0 | failed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.0 | passed |
| Single-node replicaset mongodb 5.0 | passed |
| Restore on new cluster mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | failed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.0 | failed |
| Non-sharded replicaset mongodb 4.2 | failed |
| Non-sharded replicaset mongodb 4.4 | failed |
| Restore on new cluster mongodb 5.0 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.0 | failed |
| Non-sharded replicaset mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.2 | failed |
| Single-node replicaset mongodb 4.4 | failed |
| Single-node replicaset mongodb 5.0 | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | failed |
| Sharded cluster mongodb 4.2 | failed |
| Sharded cluster mongodb 5.0 | failed |
| Sharded cluster mongodb 4.4 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.0 | failed |
| Non-sharded replicaset mongodb 4.2 | failed |
| Non-sharded replicaset mongodb 4.4 | failed |
| Single-node replicaset mongodb 4.0 | failed |
| Restore on new cluster mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |
| Restore on new cluster mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.2 | failed |
| Single-node replicaset mongodb 4.4 | failed |
| Single-node replicaset mongodb 5.0 | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | passed |
| Sharded cluster mongodb 4.4 | passed |
| Sharded cluster mongodb 5.0 | failed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Single-node replicaset mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 4.4 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | passed |
| Sharded cluster mongodb 4.4 | passed |
| Sharded cluster mongodb 5.0 | failed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.0 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | passed |
| Single-node replicaset mongodb 4.2 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |
| Restore on new cluster mongodb 5.0 | passed |

File: e2e-tests/pkg/pbm/pbm_ctl.go [hunk 1]
```diff
@@ -271,6 +271,11 @@ func (c *Ctl) PITRestore(t time.Time) error {
 	return err
 }
 
+func (c *Ctl) PITRestoreClusterTime(t, i uint32) error {
+	_, err := c.RunCmd("pbm", "restore", "--time", fmt.Sprintf("%d,%d", t, i))
```
Comment (defbin) [inline e2e-tests/pkg/pbm/pbm_ctl.go:275]: Can we use a separate flag for cluster time? it simplifies parsing and makes it obvious in usage and implementation

File: cli/restore.go [hunk 2]
```diff
@@ -107,10 +110,32 @@ func restore(cn *pbm.PBM, bcpName string, outf outFormat) error {
 	return waitForRestoreStatus(ctx, cn, name)
 }
 
-func pitrestore(cn *pbm.PBM, t, base string, outf outFormat) error {
+func parseTS(t string) (ts primitive.Timestamp, err error) {
+	if si := strings.Index(t, ","); si != -1 {
```
Comment (defbin) [inline cli/restore.go:114]: can we use [`strings.SplitN()`](https://pkg.go.dev/strings#SplitN)? the code makes understanding hard and not much optimization in runtime

File: pbm/restore.go [hunk 3]
```diff
@@ -29,12 +30,53 @@ type RestoreReplset struct {
...
+const (
+	TxnCommit  TxnState = "commit"
+	TxnPrepare TxnState = "prepare"
+	TxnAbort   TxnState = "abort"
+	TxnUnknown TxnState = ""
+)
+
+type RestoreTxn struct {
+	ID    string              `bson:"id" json:"id"`
+	Ctime primitive.Timestamp `bson:"ts" json:"ts"`
```
Comment (defbin) [inline pbm/restore.go:52]: what does `Ctime` mean? please leave a comment about it

**Review by defbin — CHANGES_REQUESTED** [review]

File: e2e-tests/pkg/pbm/pbm_ctl.go (hunk 1 above)
Comment (dAdAbird) [reply e2e-tests/pkg/pbm/pbm_ctl.go:275]: I'd leave it like this for now. I don't want to expose this functionality make any UI commitments on this stage

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| Sharded cluster mongodb 4.0 | passed |
| Sharded cluster mongodb 4.2 | passed |
| Sharded cluster mongodb 4.4 | passed |
| Sharded cluster mongodb 5.0 | failed |
| Restore on new cluster mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.2 | passed |
| Restore on new cluster mongodb 4.4 | passed |
| Non-sharded replicaset mongodb 4.0 | passed |
| Non-sharded replicaset mongodb 4.4 | passed |
| Restore on new cluster mongodb 5.0 | passed |
| Non-sharded replicaset mongodb 5.0 | failed |
| Single-node replicaset mongodb 4.2 | passed |
| Single-node replicaset mongodb 4.0 | passed |
| Single-node replicaset mongodb 4.4 | passed |
| Single-node replicaset mongodb 5.0 | passed |

**Pull request merged**